### ISSUE TITLE
Feature Request: Employ Standard Unix Input Convention for `--reads`

### ISSUE THREAD

**Issue by cbird808** [mhoban/rainbow_bridge#81]
Hi, great package!

I have a request, or I'd be happy to do the work and send you pull req if you'd implement these.

1. In the unix philosophy, every command accepts input file(s) rather than directories.  Following this, when calling `--paired`, I suggest making `--reads`, `--fwd`, and `--rev` accept a pattern (glob) to match file names rather than a directory name, and you [hardcoding the file path and naming convention](https://github.com/mhoban/rainbow_bridge?tab=readme-ov-file#processing-fastq-input-demultiplexed-or-not). It's trivial to change this and would also make the behavior of `--reads` match not only expectation based on all unix commands, but also its behavior when `--single` is set. It would also allow you to divorce from locking the end user into having `fastq` in the file name. `fq` is also common.

2. With the previous request implemented, you can extract the FASTQ file extension from the argument given for `--reads`, `--fwd`, and `--rev`

With these changes, the end users are freed from having to rename their files to match the expectations of `rainbow_bridge`.

**Comment by mhoban** [comment]
This is a perfectly reasonable request. I think there was a reason I didn't do it this way in the first place, but I can't think of why that was. It may have had partially to do with the fact that inputs could be any of single or multiple forward and/or reverse reads, depending on the (de)multiplexing strateg(ies, see e.g. #82 ) used for the sequencing run. 

I firmly agree that users shouldn't have to rename their files to match internal expectations of the pipeline. 

I'd welcome the pull request, but any implementation will need to allow the multiple input scenarios.

**Comment by cbird808** [comment]
Cool.  I'm speaking from experience here.  My intro to bash was forking ddocent 10 years ago.  After I gained experience, I realized one original sin of that software is having an inflexible file naming convention and not following unix conventions. It causes issues to this day.

Here's my thoughts on how to make the changes:


  1.  You probably set the default --reads based on either --single or --paired being called. The default for -paired -reads changes from "." To "./*.f*q*"
     *   The code will need to be searched for usage of $reads and updated as necessary to comply with a file path glob rather than a dir
  2.  For -fwd and -rev (note my computer autocorrects -- to -), the defaults are "./*${r1}.f*q.*" and "./*${r2}.f*q*" .   if they are user-specified then -reads is ignored.
     *   The code will need to be searched for usage of $fwd and $rev and updated as necessary to comply with a file path glob rather than a dir
  3.  The present code used to piece together the file paths from -reads, --fwd, and -rev will be deprecated/ commented out
     *   Any dependencies on variables created here will need to be rectified
  4.  The fasta extension will be stored in a variable and extracted using `FASTQ=$(basename "$reads" | sed -E 's/.*\.([^.]*(\.gz)?)$/\1/')`,  or equivalent for $fwd $rev
     *   Every instance of `fastq` will need to be replaced with the variable containing the file extension

Any thoughts on this?

**Comment by cbird808** [comment]
Hi @mhoban ,

I've added glob handling functionality for `--reads`, `--fwd`, and `--rev` while maintaining the original functionality described in the manual.  Use the attached files in place of the orig (brief `README.md` included). I've tested with the `test_paired_demuxed.sh`. By my reading of the code, the other tests shouldn't be affected.  

As is, I think it will handle any combination of globs from `--fwd` and `--rev`.  It will also handle globs from `-reads` that follow the input rules for a glob passed to `Channel.fromFilePairs`.  

If I were to make one primary change to what I did, it would be to take the verbose `NextFlow` code used create a glob that's accepted by `Channel.fromFilePairs` from `params.fwd` and `params.rev` and  rewrite it in pure `groovy`.  Because i was not familiar with `nextflow`, I learned the hard way that making the present `nextflow` code a process is a no-go because a channel, which is output by a process, cannot be converted back to a string, which is the only format accepted by `Channel.fromFilePairs`.

That being said, the core premise of the code I added, identifying the characters that are(nt) in common between strings, could be repurposed and modified to handle any `--reads` glob, as well as `--fwd` and `--rev` dirs without `--reads`.

Let me know how it works for you.
[ceb_update_1.zip](https://github.com/user-attachments/files/16683236/ceb_update_1.zip)

**Comment by mhoban** [comment]
Thanks for this! I'm in the process of working on some other updates but I will take a look at this and merge in your modifications once those are implemented. I appreciate that you have retained the existing functionality, since things were largely designed as they were for ease of use by less unix-savvy users (viz. grad students in the tobo lab)

**Comment by mhoban** [comment]
Thanks for you contribution, @cbird808! This is now all implemented in the branch [81-command-line](https://github.com/mhoban/rainbow_bridge/tree/81-command-line) and will soon be merged into main. I've also updated all the documentation. Major code changes can be found in 522ae01 and doc updates are in cbbf522.

I changed some stuff from your code* but the functionality is all more or less the same. The pipeline should now support arbitrary file extensions through your updates, and it has been updated to look for '\*.f\*q\*' if doing things the old way (so at least .fq, .fastq, .fq.gz, and .fastq.gz will work). 

I'll also update the test repository in a branch of the same name (but it will probably be merged soon).

*notably your common-prefix code. Your implementation was very clever, but there was an issue whereby indexOf wasn't guaranteed to match the running index in the takeWhile closure. I updated it to use a simple for-loop and it resolved the problems I was having.
